hashicorp/consul
# [CC-7049] Stop the HCP manager when link is deleted

### Description

Depends on https://github.com/hashicorp/consul/pull/20312

Stops the connection to HCP when the HCP link is deleted by:

1. Adding a stop method to the telemetry provider
2. Adding a stop method to the HCP manager that stops the SCADA provider, stops the telemetry provider, and exists update status loop.
3. Adding a token deletion method to the HCP manager and calling it from the HCP manager stop method.
4. Calling the HCP manager stop method from the finalizer and when the link is not found by the reconciler.

This PR also adds interfaces for the HCP manager and the telemetry provider to make testing easier.

### Testing & Reproduction steps

1. Create an HCP link and verify it is connected to HCP Consul Central
2. Make a DELETE request to delete the link
3. Inspect the logs to see that the HCP manager has stopped and that no further updates to HCP Consul Central are occurring, like posts to the telemetry metrics endpoint.
4. Check that the token has been deleted
5. Check in HCP Consul Central that the cluster state is disconnected.
6. Create a new HCP link and verify that it connects properly.

```
2024-01-26T16:07:25.235-0600 [DEBUG] agent.server.controller-runtime: stopping HCP manager: controller=consul.io/hcp/link managed_type=hcp.v2.Link resource-id="uid:\"01HN3WYZPYHGKFHJN7HE2SS4F8\"  name:\"global\"  type:{group:\"hcp\"  group_version:\"v2\"  kind:\"Link\"}  tenancy:{peer_name:\"local\"}"
2024-01-26T16:07:25.235-0600 [INFO]  agent.hcp_manager: HCP manager stopping
2024-01-26T16:07:25.235-0600 [INFO]  agent.hcp.scada.scada-provider: stopping
2024-01-26T16:07:25.235-0600 [DEBUG] agent.hcp: telemetry config provider stopped
2024-01-26T16:07:25.235-0600 [DEBUG] agent.hcp.scada.scada-provider: backing off: seconds=10.378947345
2024-01-26T16:07:25.285-0600 [INFO]  agent.server: deleted ACL token: description="HCP Management Token"
2024-01-26T16:07:25.285-0600 [INFO]  agent.hcp_manager: HCP manager stopped
2024-01-26T16:07:25.285-0600 [DEBUG] agent.server.controller-runtime: deleting hcp-config dir: controller=consul.io/hcp/link managed_type=hcp.v2.Link resource-id="uid:\"01HN3WYZPYHGKFHJN7HE2SS4F8\"  name:\"global\"  type:{group:\"hcp\"  group_version:\"v2\"  kind:\"Link\"}  tenancy:{peer_name:\"local\"}" dir=./data-dir/hcp-config

```

### Links
- https://hashicorp.atlassian.net/browse/CC-7049

### PR Checklist
